// livekit-client/dist/room/track/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScreenSharePresets = exports.VideoPresets43 = exports.VideoPresets = exports.AudioPresets = exports.VideoPreset = void 0;
class VideoPreset {
    constructor(width, height, maxBitrate, maxFramerate) {
        this.width = width;
        this.height = height;
        this.encoding = {
            maxBitrate,
            maxFramerate,
        };
    }
    get resolution() {
        return {
            width: this.width,
            height: this.height,
            frameRate: this.encoding.maxFramerate,
            aspectRatio: this.width / this.height,
        };
    }
}
exports.VideoPreset = VideoPreset;
var AudioPresets;
(function (AudioPresets) {
    AudioPresets.telephone = {
        maxBitrate: 12000,
    };
    AudioPresets.speech = {
        maxBitrate: 20000,
    };
    AudioPresets.music = {
        maxBitrate: 32000,
    };
})(AudioPresets = exports.AudioPresets || (exports.AudioPresets = {}));
/**
 * Sane presets for video resolution/encoding
 */
exports.VideoPresets = {
    h90: new VideoPreset(160, 90, 60000, 15),
    h180: new VideoPreset(320, 180, 120000, 15),
    h216: new VideoPreset(384, 216, 180000, 15),
    h360: new VideoPreset(640, 360, 300000, 20),
    h540: new VideoPreset(960, 540, 600000, 25),
    h720: new VideoPreset(1280, 720, 2000000, 30),
    h1080: new VideoPreset(1920, 1080, 3000000, 30),
    h1440: new VideoPreset(2560, 1440, 5000000, 30),
    h2160: new VideoPreset(3840, 2160, 8000000, 30),
    /** @deprecated */
    qvga: new VideoPreset(320, 180, 120000, 10),
    /** @deprecated */
    vga: new VideoPreset(640, 360, 300000, 20),
    /** @deprecated */
    qhd: new VideoPreset(960, 540, 600000, 25),
    /** @deprecated */
    hd: new VideoPreset(1280, 720, 2000000, 30),
    /** @deprecated */
    fhd: new VideoPreset(1920, 1080, 3000000, 30),
};
/**
 * Four by three presets
 */
exports.VideoPresets43 = {
    h120: new VideoPreset(160, 120, 80000, 15),
    h180: new VideoPreset(240, 180, 100000, 15),
    h240: new VideoPreset(320, 240, 150000, 15),
    h360: new VideoPreset(480, 360, 225000, 20),
    h480: new VideoPreset(640, 480, 300000, 20),
    h540: new VideoPreset(720, 540, 450000, 25),
    h720: new VideoPreset(960, 720, 1500000, 30),
    h1080: new VideoPreset(1440, 1080, 2500000, 30),
    h1440: new VideoPreset(1920, 1440, 3500000, 30),
    /** @deprecated */
    qvga: new VideoPreset(240, 180, 90000, 10),
    /** @deprecated */
    vga: new VideoPreset(480, 360, 225000, 20),
    /** @deprecated */
    qhd: new VideoPreset(720, 540, 450000, 25),
    /** @deprecated */
    hd: new VideoPreset(960, 720, 1500000, 30),
    /** @deprecated */
    fhd: new VideoPreset(1440, 1080, 2800000, 30),
};
exports.ScreenSharePresets = {
    h360fps3: new VideoPreset(640, 360, 200000, 3),
    h720fps5: new VideoPreset(1280, 720, 400000, 5),
    h720fps15: new VideoPreset(1280, 720, 1000000, 15),
    h1080fps15: new VideoPreset(1920, 1080, 1500000, 15),
    h1080fps30: new VideoPreset(1920, 1080, 3000000, 30),
    /** @deprecated */
    vga: new VideoPreset(640, 360, 200000, 3),
    /** @deprecated */
    hd_8: new VideoPreset(1280, 720, 400000, 5),
    /** @deprecated */
    hd_15: new VideoPreset(1280, 720, 1000000, 15),
    /** @deprecated */
    fhd_15: new VideoPreset(1920, 1080, 1500000, 15),
    /** @deprecated */
    fhd_30: new VideoPreset(1920, 1080, 3000000, 30),
};
