// livekit-client/dist/room/track/TrackPublication.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrackPublication = void 0;
const events_1 = require("events");
const events_2 = require("../events");
const LocalAudioTrack_1 = __importDefault(require("./LocalAudioTrack"));
const LocalVideoTrack_1 = __importDefault(require("./LocalVideoTrack"));
const RemoteAudioTrack_1 = __importDefault(require("./RemoteAudioTrack"));
const RemoteVideoTrack_1 = __importDefault(require("./RemoteVideoTrack"));
const Track_1 = require("./Track");
class TrackPublication extends events_1.EventEmitter {
    constructor(kind, id, name) {
        super();
        this.metadataMuted = false;
        this.handleMuted = () => {
            this.emit(events_2.TrackEvent.Muted);
        };
        this.handleUnmuted = () => {
            this.emit(events_2.TrackEvent.Unmuted);
        };
        this.kind = kind;
        this.trackSid = id;
        this.trackName = name;
        this.source = Track_1.Track.Source.Unknown;
    }
    /** @internal */
    setTrack(track) {
        if (this.track) {
            this.track.off(events_2.TrackEvent.Muted, this.handleMuted);
            this.track.off(events_2.TrackEvent.Unmuted, this.handleUnmuted);
        }
        this.track = track;
        if (track) {
            // forward events
            track.on(events_2.TrackEvent.Muted, this.handleMuted);
            track.on(events_2.TrackEvent.Unmuted, this.handleUnmuted);
        }
    }
    get isMuted() {
        return this.metadataMuted;
    }
    get isEnabled() {
        return true;
    }
    get isSubscribed() {
        return this.track !== undefined;
    }
    /**
     * an [AudioTrack] if this publication holds an audio track
     */
    get audioTrack() {
        if (this.track instanceof LocalAudioTrack_1.default || this.track instanceof RemoteAudioTrack_1.default) {
            return this.track;
        }
    }
    /**
     * an [VideoTrack] if this publication holds a video track
     */
    get videoTrack() {
        if (this.track instanceof LocalVideoTrack_1.default || this.track instanceof RemoteVideoTrack_1.default) {
            return this.track;
        }
    }
    /** @internal */
    updateInfo(info) {
        this.trackSid = info.sid;
        this.trackName = info.name;
        this.source = Track_1.Track.sourceFromProto(info.source);
        this.mimeType = info.mimeType;
        if (this.kind === Track_1.Track.Kind.Video && info.width > 0) {
            this.dimensions = {
                width: info.width,
                height: info.height,
            };
            this.simulcasted = info.simulcast;
        }
        this.trackInfo = info;
    }
}
exports.TrackPublication = TrackPublication;
(function (TrackPublication) {
    let SubscriptionStatus;
    (function (SubscriptionStatus) {
        SubscriptionStatus["Subscribed"] = "subscribed";
        SubscriptionStatus["NotAllowed"] = "not_allowed";
        SubscriptionStatus["Unsubscribed"] = "unsubscribed";
    })(SubscriptionStatus = TrackPublication.SubscriptionStatus || (TrackPublication.SubscriptionStatus = {}));
})(TrackPublication = exports.TrackPublication || (exports.TrackPublication = {}));
