// livekit-client/dist/proto/livekit_rtc.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimulateScenario = exports.DataChannelInfo = exports.SyncState = exports.SubscriptionPermissionUpdate = exports.SubscriptionPermission = exports.TrackPermission = exports.SubscribedQualityUpdate = exports.SubscribedQuality = exports.StreamStateUpdate = exports.StreamStateInfo = exports.ConnectionQualityUpdate = exports.ConnectionQualityInfo = exports.RoomUpdate = exports.SpeakersChanged = exports.ICEServer = exports.UpdateVideoLayers = exports.LeaveRequest = exports.UpdateTrackSettings = exports.UpdateSubscription = exports.ParticipantUpdate = exports.SessionDescription = exports.TrackUnpublishedResponse = exports.TrackPublishedResponse = exports.JoinResponse = exports.MuteTrackRequest = exports.TrickleRequest = exports.AddTrackRequest = exports.SignalResponse = exports.SignalRequest = exports.streamStateToJSON = exports.streamStateFromJSON = exports.StreamState = exports.signalTargetToJSON = exports.signalTargetFromJSON = exports.SignalTarget = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const _m0 = __importStar(require("protobufjs/minimal"));
const livekit_models_1 = require("./livekit_models");
exports.protobufPackage = "livekit";
var SignalTarget;
(function (SignalTarget) {
    SignalTarget[SignalTarget["PUBLISHER"] = 0] = "PUBLISHER";
    SignalTarget[SignalTarget["SUBSCRIBER"] = 1] = "SUBSCRIBER";
    SignalTarget[SignalTarget["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(SignalTarget = exports.SignalTarget || (exports.SignalTarget = {}));
function signalTargetFromJSON(object) {
    switch (object) {
        case 0:
        case "PUBLISHER":
            return SignalTarget.PUBLISHER;
        case 1:
        case "SUBSCRIBER":
            return SignalTarget.SUBSCRIBER;
        case -1:
        case "UNRECOGNIZED":
        default:
            return SignalTarget.UNRECOGNIZED;
    }
}
exports.signalTargetFromJSON = signalTargetFromJSON;
function signalTargetToJSON(object) {
    switch (object) {
        case SignalTarget.PUBLISHER:
            return "PUBLISHER";
        case SignalTarget.SUBSCRIBER:
            return "SUBSCRIBER";
        default:
            return "UNKNOWN";
    }
}
exports.signalTargetToJSON = signalTargetToJSON;
var StreamState;
(function (StreamState) {
    StreamState[StreamState["ACTIVE"] = 0] = "ACTIVE";
    StreamState[StreamState["PAUSED"] = 1] = "PAUSED";
    StreamState[StreamState["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(StreamState = exports.StreamState || (exports.StreamState = {}));
function streamStateFromJSON(object) {
    switch (object) {
        case 0:
        case "ACTIVE":
            return StreamState.ACTIVE;
        case 1:
        case "PAUSED":
            return StreamState.PAUSED;
        case -1:
        case "UNRECOGNIZED":
        default:
            return StreamState.UNRECOGNIZED;
    }
}
exports.streamStateFromJSON = streamStateFromJSON;
function streamStateToJSON(object) {
    switch (object) {
        case StreamState.ACTIVE:
            return "ACTIVE";
        case StreamState.PAUSED:
            return "PAUSED";
        default:
            return "UNKNOWN";
    }
}
exports.streamStateToJSON = streamStateToJSON;
function createBaseSignalRequest() {
    return {
        offer: undefined,
        answer: undefined,
        trickle: undefined,
        addTrack: undefined,
        mute: undefined,
        subscription: undefined,
        trackSetting: undefined,
        leave: undefined,
        updateLayers: undefined,
        subscriptionPermission: undefined,
        syncState: undefined,
        simulate: undefined,
    };
}
exports.SignalRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.offer !== undefined) {
            exports.SessionDescription.encode(message.offer, writer.uint32(10).fork()).ldelim();
        }
        if (message.answer !== undefined) {
            exports.SessionDescription.encode(message.answer, writer.uint32(18).fork()).ldelim();
        }
        if (message.trickle !== undefined) {
            exports.TrickleRequest.encode(message.trickle, writer.uint32(26).fork()).ldelim();
        }
        if (message.addTrack !== undefined) {
            exports.AddTrackRequest.encode(message.addTrack, writer.uint32(34).fork()).ldelim();
        }
        if (message.mute !== undefined) {
            exports.MuteTrackRequest.encode(message.mute, writer.uint32(42).fork()).ldelim();
        }
        if (message.subscription !== undefined) {
            exports.UpdateSubscription.encode(message.subscription, writer.uint32(50).fork()).ldelim();
        }
        if (message.trackSetting !== undefined) {
            exports.UpdateTrackSettings.encode(message.trackSetting, writer.uint32(58).fork()).ldelim();
        }
        if (message.leave !== undefined) {
            exports.LeaveRequest.encode(message.leave, writer.uint32(66).fork()).ldelim();
        }
        if (message.updateLayers !== undefined) {
            exports.UpdateVideoLayers.encode(message.updateLayers, writer.uint32(82).fork()).ldelim();
        }
        if (message.subscriptionPermission !== undefined) {
            exports.SubscriptionPermission.encode(message.subscriptionPermission, writer.uint32(90).fork()).ldelim();
        }
        if (message.syncState !== undefined) {
            exports.SyncState.encode(message.syncState, writer.uint32(98).fork()).ldelim();
        }
        if (message.simulate !== undefined) {
            exports.SimulateScenario.encode(message.simulate, writer.uint32(106).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSignalRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.offer = exports.SessionDescription.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.answer = exports.SessionDescription.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.trickle = exports.TrickleRequest.decode(reader, reader.uint32());
                    break;
                case 4:
                    message.addTrack = exports.AddTrackRequest.decode(reader, reader.uint32());
                    break;
                case 5:
                    message.mute = exports.MuteTrackRequest.decode(reader, reader.uint32());
                    break;
                case 6:
                    message.subscription = exports.UpdateSubscription.decode(reader, reader.uint32());
                    break;
                case 7:
                    message.trackSetting = exports.UpdateTrackSettings.decode(reader, reader.uint32());
                    break;
                case 8:
                    message.leave = exports.LeaveRequest.decode(reader, reader.uint32());
                    break;
                case 10:
                    message.updateLayers = exports.UpdateVideoLayers.decode(reader, reader.uint32());
                    break;
                case 11:
                    message.subscriptionPermission = exports.SubscriptionPermission.decode(reader, reader.uint32());
                    break;
                case 12:
                    message.syncState = exports.SyncState.decode(reader, reader.uint32());
                    break;
                case 13:
                    message.simulate = exports.SimulateScenario.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            offer: isSet(object.offer)
                ? exports.SessionDescription.fromJSON(object.offer)
                : undefined,
            answer: isSet(object.answer)
                ? exports.SessionDescription.fromJSON(object.answer)
                : undefined,
            trickle: isSet(object.trickle)
                ? exports.TrickleRequest.fromJSON(object.trickle)
                : undefined,
            addTrack: isSet(object.addTrack)
                ? exports.AddTrackRequest.fromJSON(object.addTrack)
                : undefined,
            mute: isSet(object.mute)
                ? exports.MuteTrackRequest.fromJSON(object.mute)
                : undefined,
            subscription: isSet(object.subscription)
                ? exports.UpdateSubscription.fromJSON(object.subscription)
                : undefined,
            trackSetting: isSet(object.trackSetting)
                ? exports.UpdateTrackSettings.fromJSON(object.trackSetting)
                : undefined,
            leave: isSet(object.leave)
                ? exports.LeaveRequest.fromJSON(object.leave)
                : undefined,
            updateLayers: isSet(object.updateLayers)
                ? exports.UpdateVideoLayers.fromJSON(object.updateLayers)
                : undefined,
            subscriptionPermission: isSet(object.subscriptionPermission)
                ? exports.SubscriptionPermission.fromJSON(object.subscriptionPermission)
                : undefined,
            syncState: isSet(object.syncState)
                ? exports.SyncState.fromJSON(object.syncState)
                : undefined,
            simulate: isSet(object.simulate)
                ? exports.SimulateScenario.fromJSON(object.simulate)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.offer !== undefined &&
            (obj.offer = message.offer
                ? exports.SessionDescription.toJSON(message.offer)
                : undefined);
        message.answer !== undefined &&
            (obj.answer = message.answer
                ? exports.SessionDescription.toJSON(message.answer)
                : undefined);
        message.trickle !== undefined &&
            (obj.trickle = message.trickle
                ? exports.TrickleRequest.toJSON(message.trickle)
                : undefined);
        message.addTrack !== undefined &&
            (obj.addTrack = message.addTrack
                ? exports.AddTrackRequest.toJSON(message.addTrack)
                : undefined);
        message.mute !== undefined &&
            (obj.mute = message.mute
                ? exports.MuteTrackRequest.toJSON(message.mute)
                : undefined);
        message.subscription !== undefined &&
            (obj.subscription = message.subscription
                ? exports.UpdateSubscription.toJSON(message.subscription)
                : undefined);
        message.trackSetting !== undefined &&
            (obj.trackSetting = message.trackSetting
                ? exports.UpdateTrackSettings.toJSON(message.trackSetting)
                : undefined);
        message.leave !== undefined &&
            (obj.leave = message.leave
                ? exports.LeaveRequest.toJSON(message.leave)
                : undefined);
        message.updateLayers !== undefined &&
            (obj.updateLayers = message.updateLayers
                ? exports.UpdateVideoLayers.toJSON(message.updateLayers)
                : undefined);
        message.subscriptionPermission !== undefined &&
            (obj.subscriptionPermission = message.subscriptionPermission
                ? exports.SubscriptionPermission.toJSON(message.subscriptionPermission)
                : undefined);
        message.syncState !== undefined &&
            (obj.syncState = message.syncState
                ? exports.SyncState.toJSON(message.syncState)
                : undefined);
        message.simulate !== undefined &&
            (obj.simulate = message.simulate
                ? exports.SimulateScenario.toJSON(message.simulate)
                : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseSignalRequest();
        message.offer =
            object.offer !== undefined && object.offer !== null
                ? exports.SessionDescription.fromPartial(object.offer)
                : undefined;
        message.answer =
            object.answer !== undefined && object.answer !== null
                ? exports.SessionDescription.fromPartial(object.answer)
                : undefined;
        message.trickle =
            object.trickle !== undefined && object.trickle !== null
                ? exports.TrickleRequest.fromPartial(object.trickle)
                : undefined;
        message.addTrack =
            object.addTrack !== undefined && object.addTrack !== null
                ? exports.AddTrackRequest.fromPartial(object.addTrack)
                : undefined;
        message.mute =
            object.mute !== undefined && object.mute !== null
                ? exports.MuteTrackRequest.fromPartial(object.mute)
                : undefined;
        message.subscription =
            object.subscription !== undefined && object.subscription !== null
                ? exports.UpdateSubscription.fromPartial(object.subscription)
                : undefined;
        message.trackSetting =
            object.trackSetting !== undefined && object.trackSetting !== null
                ? exports.UpdateTrackSettings.fromPartial(object.trackSetting)
                : undefined;
        message.leave =
            object.leave !== undefined && object.leave !== null
                ? exports.LeaveRequest.fromPartial(object.leave)
                : undefined;
        message.updateLayers =
            object.updateLayers !== undefined && object.updateLayers !== null
                ? exports.UpdateVideoLayers.fromPartial(object.updateLayers)
                : undefined;
        message.subscriptionPermission =
            object.subscriptionPermission !== undefined &&
                object.subscriptionPermission !== null
                ? exports.SubscriptionPermission.fromPartial(object.subscriptionPermission)
                : undefined;
        message.syncState =
            object.syncState !== undefined && object.syncState !== null
                ? exports.SyncState.fromPartial(object.syncState)
                : undefined;
        message.simulate =
            object.simulate !== undefined && object.simulate !== null
                ? exports.SimulateScenario.fromPartial(object.simulate)
                : undefined;
        return message;
    },
};
function createBaseSignalResponse() {
    return {
        join: undefined,
        answer: undefined,
        offer: undefined,
        trickle: undefined,
        update: undefined,
        trackPublished: undefined,
        leave: undefined,
        mute: undefined,
        speakersChanged: undefined,
        roomUpdate: undefined,
        connectionQuality: undefined,
        streamStateUpdate: undefined,
        subscribedQualityUpdate: undefined,
        subscriptionPermissionUpdate: undefined,
        refreshToken: undefined,
        trackUnpublished: undefined,
    };
}
exports.SignalResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.join !== undefined) {
            exports.JoinResponse.encode(message.join, writer.uint32(10).fork()).ldelim();
        }
        if (message.answer !== undefined) {
            exports.SessionDescription.encode(message.answer, writer.uint32(18).fork()).ldelim();
        }
        if (message.offer !== undefined) {
            exports.SessionDescription.encode(message.offer, writer.uint32(26).fork()).ldelim();
        }
        if (message.trickle !== undefined) {
            exports.TrickleRequest.encode(message.trickle, writer.uint32(34).fork()).ldelim();
        }
        if (message.update !== undefined) {
            exports.ParticipantUpdate.encode(message.update, writer.uint32(42).fork()).ldelim();
        }
        if (message.trackPublished !== undefined) {
            exports.TrackPublishedResponse.encode(message.trackPublished, writer.uint32(50).fork()).ldelim();
        }
        if (message.leave !== undefined) {
            exports.LeaveRequest.encode(message.leave, writer.uint32(66).fork()).ldelim();
        }
        if (message.mute !== undefined) {
            exports.MuteTrackRequest.encode(message.mute, writer.uint32(74).fork()).ldelim();
        }
        if (message.speakersChanged !== undefined) {
            exports.SpeakersChanged.encode(message.speakersChanged, writer.uint32(82).fork()).ldelim();
        }
        if (message.roomUpdate !== undefined) {
            exports.RoomUpdate.encode(message.roomUpdate, writer.uint32(90).fork()).ldelim();
        }
        if (message.connectionQuality !== undefined) {
            exports.ConnectionQualityUpdate.encode(message.connectionQuality, writer.uint32(98).fork()).ldelim();
        }
        if (message.streamStateUpdate !== undefined) {
            exports.StreamStateUpdate.encode(message.streamStateUpdate, writer.uint32(106).fork()).ldelim();
        }
        if (message.subscribedQualityUpdate !== undefined) {
            exports.SubscribedQualityUpdate.encode(message.subscribedQualityUpdate, writer.uint32(114).fork()).ldelim();
        }
        if (message.subscriptionPermissionUpdate !== undefined) {
            exports.SubscriptionPermissionUpdate.encode(message.subscriptionPermissionUpdate, writer.uint32(122).fork()).ldelim();
        }
        if (message.refreshToken !== undefined) {
            writer.uint32(130).string(message.refreshToken);
        }
        if (message.trackUnpublished !== undefined) {
            exports.TrackUnpublishedResponse.encode(message.trackUnpublished, writer.uint32(138).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSignalResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.join = exports.JoinResponse.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.answer = exports.SessionDescription.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.offer = exports.SessionDescription.decode(reader, reader.uint32());
                    break;
                case 4:
                    message.trickle = exports.TrickleRequest.decode(reader, reader.uint32());
                    break;
                case 5:
                    message.update = exports.ParticipantUpdate.decode(reader, reader.uint32());
                    break;
                case 6:
                    message.trackPublished = exports.TrackPublishedResponse.decode(reader, reader.uint32());
                    break;
                case 8:
                    message.leave = exports.LeaveRequest.decode(reader, reader.uint32());
                    break;
                case 9:
                    message.mute = exports.MuteTrackRequest.decode(reader, reader.uint32());
                    break;
                case 10:
                    message.speakersChanged = exports.SpeakersChanged.decode(reader, reader.uint32());
                    break;
                case 11:
                    message.roomUpdate = exports.RoomUpdate.decode(reader, reader.uint32());
                    break;
                case 12:
                    message.connectionQuality = exports.ConnectionQualityUpdate.decode(reader, reader.uint32());
                    break;
                case 13:
                    message.streamStateUpdate = exports.StreamStateUpdate.decode(reader, reader.uint32());
                    break;
                case 14:
                    message.subscribedQualityUpdate = exports.SubscribedQualityUpdate.decode(reader, reader.uint32());
                    break;
                case 15:
                    message.subscriptionPermissionUpdate =
                        exports.SubscriptionPermissionUpdate.decode(reader, reader.uint32());
                    break;
                case 16:
                    message.refreshToken = reader.string();
                    break;
                case 17:
                    message.trackUnpublished = exports.TrackUnpublishedResponse.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            join: isSet(object.join) ? exports.JoinResponse.fromJSON(object.join) : undefined,
            answer: isSet(object.answer)
                ? exports.SessionDescription.fromJSON(object.answer)
                : undefined,
            offer: isSet(object.offer)
                ? exports.SessionDescription.fromJSON(object.offer)
                : undefined,
            trickle: isSet(object.trickle)
                ? exports.TrickleRequest.fromJSON(object.trickle)
                : undefined,
            update: isSet(object.update)
                ? exports.ParticipantUpdate.fromJSON(object.update)
                : undefined,
            trackPublished: isSet(object.trackPublished)
                ? exports.TrackPublishedResponse.fromJSON(object.trackPublished)
                : undefined,
            leave: isSet(object.leave)
                ? exports.LeaveRequest.fromJSON(object.leave)
                : undefined,
            mute: isSet(object.mute)
                ? exports.MuteTrackRequest.fromJSON(object.mute)
                : undefined,
            speakersChanged: isSet(object.speakersChanged)
                ? exports.SpeakersChanged.fromJSON(object.speakersChanged)
                : undefined,
            roomUpdate: isSet(object.roomUpdate)
                ? exports.RoomUpdate.fromJSON(object.roomUpdate)
                : undefined,
            connectionQuality: isSet(object.connectionQuality)
                ? exports.ConnectionQualityUpdate.fromJSON(object.connectionQuality)
                : undefined,
            streamStateUpdate: isSet(object.streamStateUpdate)
                ? exports.StreamStateUpdate.fromJSON(object.streamStateUpdate)
                : undefined,
            subscribedQualityUpdate: isSet(object.subscribedQualityUpdate)
                ? exports.SubscribedQualityUpdate.fromJSON(object.subscribedQualityUpdate)
                : undefined,
            subscriptionPermissionUpdate: isSet(object.subscriptionPermissionUpdate)
                ? exports.SubscriptionPermissionUpdate.fromJSON(object.subscriptionPermissionUpdate)
                : undefined,
            refreshToken: isSet(object.refreshToken)
                ? String(object.refreshToken)
                : undefined,
            trackUnpublished: isSet(object.trackUnpublished)
                ? exports.TrackUnpublishedResponse.fromJSON(object.trackUnpublished)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.join !== undefined &&
            (obj.join = message.join ? exports.JoinResponse.toJSON(message.join) : undefined);
        message.answer !== undefined &&
            (obj.answer = message.answer
                ? exports.SessionDescription.toJSON(message.answer)
                : undefined);
        message.offer !== undefined &&
            (obj.offer = message.offer
                ? exports.SessionDescription.toJSON(message.offer)
                : undefined);
        message.trickle !== undefined &&
            (obj.trickle = message.trickle
                ? exports.TrickleRequest.toJSON(message.trickle)
                : undefined);
        message.update !== undefined &&
            (obj.update = message.update
                ? exports.ParticipantUpdate.toJSON(message.update)
                : undefined);
        message.trackPublished !== undefined &&
            (obj.trackPublished = message.trackPublished
                ? exports.TrackPublishedResponse.toJSON(message.trackPublished)
                : undefined);
        message.leave !== undefined &&
            (obj.leave = message.leave
                ? exports.LeaveRequest.toJSON(message.leave)
                : undefined);
        message.mute !== undefined &&
            (obj.mute = message.mute
                ? exports.MuteTrackRequest.toJSON(message.mute)
                : undefined);
        message.speakersChanged !== undefined &&
            (obj.speakersChanged = message.speakersChanged
                ? exports.SpeakersChanged.toJSON(message.speakersChanged)
                : undefined);
        message.roomUpdate !== undefined &&
            (obj.roomUpdate = message.roomUpdate
                ? exports.RoomUpdate.toJSON(message.roomUpdate)
                : undefined);
        message.connectionQuality !== undefined &&
            (obj.connectionQuality = message.connectionQuality
                ? exports.ConnectionQualityUpdate.toJSON(message.connectionQuality)
                : undefined);
        message.streamStateUpdate !== undefined &&
            (obj.streamStateUpdate = message.streamStateUpdate
                ? exports.StreamStateUpdate.toJSON(message.streamStateUpdate)
                : undefined);
        message.subscribedQualityUpdate !== undefined &&
            (obj.subscribedQualityUpdate = message.subscribedQualityUpdate
                ? exports.SubscribedQualityUpdate.toJSON(message.subscribedQualityUpdate)
                : undefined);
        message.subscriptionPermissionUpdate !== undefined &&
            (obj.subscriptionPermissionUpdate = message.subscriptionPermissionUpdate
                ? exports.SubscriptionPermissionUpdate.toJSON(message.subscriptionPermissionUpdate)
                : undefined);
        message.refreshToken !== undefined &&
            (obj.refreshToken = message.refreshToken);
        message.trackUnpublished !== undefined &&
            (obj.trackUnpublished = message.trackUnpublished
                ? exports.TrackUnpublishedResponse.toJSON(message.trackUnpublished)
                : undefined);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseSignalResponse();
        message.join =
            object.join !== undefined && object.join !== null
                ? exports.JoinResponse.fromPartial(object.join)
                : undefined;
        message.answer =
            object.answer !== undefined && object.answer !== null
                ? exports.SessionDescription.fromPartial(object.answer)
                : undefined;
        message.offer =
            object.offer !== undefined && object.offer !== null
                ? exports.SessionDescription.fromPartial(object.offer)
                : undefined;
        message.trickle =
            object.trickle !== undefined && object.trickle !== null
                ? exports.TrickleRequest.fromPartial(object.trickle)
                : undefined;
        message.update =
            object.update !== undefined && object.update !== null
                ? exports.ParticipantUpdate.fromPartial(object.update)
                : undefined;
        message.trackPublished =
            object.trackPublished !== undefined && object.trackPublished !== null
                ? exports.TrackPublishedResponse.fromPartial(object.trackPublished)
                : undefined;
        message.leave =
            object.leave !== undefined && object.leave !== null
                ? exports.LeaveRequest.fromPartial(object.leave)
                : undefined;
        message.mute =
            object.mute !== undefined && object.mute !== null
                ? exports.MuteTrackRequest.fromPartial(object.mute)
                : undefined;
        message.speakersChanged =
            object.speakersChanged !== undefined && object.speakersChanged !== null
                ? exports.SpeakersChanged.fromPartial(object.speakersChanged)
                : undefined;
        message.roomUpdate =
            object.roomUpdate !== undefined && object.roomUpdate !== null
                ? exports.RoomUpdate.fromPartial(object.roomUpdate)
                : undefined;
        message.connectionQuality =
            object.connectionQuality !== undefined &&
                object.connectionQuality !== null
                ? exports.ConnectionQualityUpdate.fromPartial(object.connectionQuality)
                : undefined;
        message.streamStateUpdate =
            object.streamStateUpdate !== undefined &&
                object.streamStateUpdate !== null
                ? exports.StreamStateUpdate.fromPartial(object.streamStateUpdate)
                : undefined;
        message.subscribedQualityUpdate =
            object.subscribedQualityUpdate !== undefined &&
                object.subscribedQualityUpdate !== null
                ? exports.SubscribedQualityUpdate.fromPartial(object.subscribedQualityUpdate)
                : undefined;
        message.subscriptionPermissionUpdate =
            object.subscriptionPermissionUpdate !== undefined &&
                object.subscriptionPermissionUpdate !== null
                ? exports.SubscriptionPermissionUpdate.fromPartial(object.subscriptionPermissionUpdate)
                : undefined;
        message.refreshToken = (_a = object.refreshToken) !== null && _a !== void 0 ? _a : undefined;
        message.trackUnpublished =
            object.trackUnpublished !== undefined && object.trackUnpublished !== null
                ? exports.TrackUnpublishedResponse.fromPartial(object.trackUnpublished)
                : undefined;
        return message;
    },
};
function createBaseAddTrackRequest() {
    return {
        cid: "",
        name: "",
        type: 0,
        width: 0,
        height: 0,
        muted: false,
        disableDtx: false,
        source: 0,
        layers: [],
    };
}
exports.AddTrackRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.cid !== "") {
            writer.uint32(10).string(message.cid);
        }
        if (message.name !== "") {
            writer.uint32(18).string(message.name);
        }
        if (message.type !== 0) {
            writer.uint32(24).int32(message.type);
        }
        if (message.width !== 0) {
            writer.uint32(32).uint32(message.width);
        }
        if (message.height !== 0) {
            writer.uint32(40).uint32(message.height);
        }
        if (message.muted === true) {
            writer.uint32(48).bool(message.muted);
        }
        if (message.disableDtx === true) {
            writer.uint32(56).bool(message.disableDtx);
        }
        if (message.source !== 0) {
            writer.uint32(64).int32(message.source);
        }
        for (const v of message.layers) {
            livekit_models_1.VideoLayer.encode(v, writer.uint32(74).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAddTrackRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.cid = reader.string();
                    break;
                case 2:
                    message.name = reader.string();
                    break;
                case 3:
                    message.type = reader.int32();
                    break;
                case 4:
                    message.width = reader.uint32();
                    break;
                case 5:
                    message.height = reader.uint32();
                    break;
                case 6:
                    message.muted = reader.bool();
                    break;
                case 7:
                    message.disableDtx = reader.bool();
                    break;
                case 8:
                    message.source = reader.int32();
                    break;
                case 9:
                    message.layers.push(livekit_models_1.VideoLayer.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            cid: isSet(object.cid) ? String(object.cid) : "",
            name: isSet(object.name) ? String(object.name) : "",
            type: isSet(object.type) ? livekit_models_1.trackTypeFromJSON(object.type) : 0,
            width: isSet(object.width) ? Number(object.width) : 0,
            height: isSet(object.height) ? Number(object.height) : 0,
            muted: isSet(object.muted) ? Boolean(object.muted) : false,
            disableDtx: isSet(object.disableDtx) ? Boolean(object.disableDtx) : false,
            source: isSet(object.source) ? livekit_models_1.trackSourceFromJSON(object.source) : 0,
            layers: Array.isArray(object === null || object === void 0 ? void 0 : object.layers)
                ? object.layers.map((e) => livekit_models_1.VideoLayer.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        message.cid !== undefined && (obj.cid = message.cid);
        message.name !== undefined && (obj.name = message.name);
        message.type !== undefined && (obj.type = livekit_models_1.trackTypeToJSON(message.type));
        message.width !== undefined && (obj.width = Math.round(message.width));
        message.height !== undefined && (obj.height = Math.round(message.height));
        message.muted !== undefined && (obj.muted = message.muted);
        message.disableDtx !== undefined && (obj.disableDtx = message.disableDtx);
        message.source !== undefined &&
            (obj.source = livekit_models_1.trackSourceToJSON(message.source));
        if (message.layers) {
            obj.layers = message.layers.map((e) => e ? livekit_models_1.VideoLayer.toJSON(e) : undefined);
        }
        else {
            obj.layers = [];
        }
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j;
        const message = createBaseAddTrackRequest();
        message.cid = (_a = object.cid) !== null && _a !== void 0 ? _a : "";
        message.name = (_b = object.name) !== null && _b !== void 0 ? _b : "";
        message.type = (_c = object.type) !== null && _c !== void 0 ? _c : 0;
        message.width = (_d = object.width) !== null && _d !== void 0 ? _d : 0;
        message.height = (_e = object.height) !== null && _e !== void 0 ? _e : 0;
        message.muted = (_f = object.muted) !== null && _f !== void 0 ? _f : false;
        message.disableDtx = (_g = object.disableDtx) !== null && _g !== void 0 ? _g : false;
        message.source = (_h = object.source) !== null && _h !== void 0 ? _h : 0;
        message.layers = ((_j = object.layers) === null || _j === void 0 ? void 0 : _j.map((e) => livekit_models_1.VideoLayer.fromPartial(e))) || [];
        return message;
    },
};
function createBaseTrickleRequest() {
    return { candidateInit: "", target: 0 };
}
exports.TrickleRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.candidateInit !== "") {
            writer.uint32(10).string(message.candidateInit);
        }
        if (message.target !== 0) {
            writer.uint32(16).int32(message.target);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTrickleRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.candidateInit = reader.string();
                    break;
                case 2:
                    message.target = reader.int32();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            candidateInit: isSet(object.candidateInit)
                ? String(object.candidateInit)
                : "",
            target: isSet(object.target) ? signalTargetFromJSON(object.target) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        message.candidateInit !== undefined &&
            (obj.candidateInit = message.candidateInit);
        message.target !== undefined &&
            (obj.target = signalTargetToJSON(message.target));
        return obj;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseTrickleRequest();
        message.candidateInit = (_a = object.candidateInit) !== null && _a !== void 0 ? _a : "";
        message.target = (_b = object.target) !== null && _b !== void 0 ? _b : 0;
        return message;
    },
};
function createBaseMuteTrackRequest() {
    return { sid: "", muted: false };
}
exports.MuteTrackRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.sid !== "") {
            writer.uint32(10).string(message.sid);
        }
        if (message.muted === true) {
            writer.uint32(16).bool(message.muted);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMuteTrackRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.sid = reader.string();
                    break;
                case 2:
                    message.muted = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            sid: isSet(object.sid) ? String(object.sid) : "",
            muted: isSet(object.muted) ? Boolean(object.muted) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        message.sid !== undefined && (obj.sid = message.sid);
        message.muted !== undefined && (obj.muted = message.muted);
        return obj;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseMuteTrackRequest();
        message.sid = (_a = object.sid) !== null && _a !== void 0 ? _a : "";
        message.muted = (_b = object.muted) !== null && _b !== void 0 ? _b : false;
        return message;
    },
};
function createBaseJoinResponse() {
    return {
        room: undefined,
        participant: undefined,
        otherParticipants: [],
        serverVersion: "",
        iceServers: [],
        subscriberPrimary: false,
        alternativeUrl: "",
        clientConfiguration: undefined,
        serverRegion: "",
    };
}
exports.JoinResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.room !== undefined) {
            livekit_models_1.Room.encode(message.room, writer.uint32(10).fork()).ldelim();
        }
        if (message.participant !== undefined) {
            livekit_models_1.ParticipantInfo.encode(message.participant, writer.uint32(18).fork()).ldelim();
        }
        for (const v of message.otherParticipants) {
            livekit_models_1.ParticipantInfo.encode(v, writer.uint32(26).fork()).ldelim();
        }
        if (message.serverVersion !== "") {
            writer.uint32(34).string(message.serverVersion);
        }
        for (const v of message.iceServers) {
            exports.ICEServer.encode(v, writer.uint32(42).fork()).ldelim();
        }
        if (message.subscriberPrimary === true) {
            writer.uint32(48).bool(message.subscriberPrimary);
        }
        if (message.alternativeUrl !== "") {
            writer.uint32(58).string(message.alternativeUrl);
        }
        if (message.clientConfiguration !== undefined) {
            livekit_models_1.ClientConfiguration.encode(message.clientConfiguration, writer.uint32(66).fork()).ldelim();
        }
        if (message.serverRegion !== "") {
            writer.uint32(74).string(message.serverRegion);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseJoinResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.room = livekit_models_1.Room.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.participant = livekit_models_1.ParticipantInfo.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.otherParticipants.push(livekit_models_1.ParticipantInfo.decode(reader, reader.uint32()));
                    break;
                case 4:
                    message.serverVersion = reader.string();
                    break;
                case 5:
                    message.iceServers.push(exports.ICEServer.decode(reader, reader.uint32()));
                    break;
                case 6:
                    message.subscriberPrimary = reader.bool();
                    break;
                case 7:
                    message.alternativeUrl = reader.string();
                    break;
                case 8:
                    message.clientConfiguration = livekit_models_1.ClientConfiguration.decode(reader, reader.uint32());
                    break;
                case 9:
                    message.serverRegion = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            room: isSet(object.room) ? livekit_models_1.Room.fromJSON(object.room) : undefined,
            participant: isSet(object.participant)
                ? livekit_models_1.ParticipantInfo.fromJSON(object.participant)
                : undefined,
            otherParticipants: Array.isArray(object === null || object === void 0 ? void 0 : object.otherParticipants)
                ? object.otherParticipants.map((e) => livekit_models_1.ParticipantInfo.fromJSON(e))
                : [],
            serverVersion: isSet(object.serverVersion)
                ? String(object.serverVersion)
                : "",
            iceServers: Array.isArray(object === null || object === void 0 ? void 0 : object.iceServers)
                ? object.iceServers.map((e) => exports.ICEServer.fromJSON(e))
                : [],
            subscriberPrimary: isSet(object.subscriberPrimary)
                ? Boolean(object.subscriberPrimary)
                : false,
            alternativeUrl: isSet(object.alternativeUrl)
                ? String(object.alternativeUrl)
                : "",
            clientConfiguration: isSet(object.clientConfiguration)
                ? livekit_models_1.ClientConfiguration.fromJSON(object.clientConfiguration)
                : undefined,
            serverRegion: isSet(object.serverRegion)
                ? String(object.serverRegion)
                : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.room !== undefined &&
            (obj.room = message.room ? livekit_models_1.Room.toJSON(message.room) : undefined);
        message.participant !== undefined &&
            (obj.participant = message.participant
                ? livekit_models_1.ParticipantInfo.toJSON(message.participant)
                : undefined);
        if (message.otherParticipants) {
            obj.otherParticipants = message.otherParticipants.map((e) => e ? livekit_models_1.ParticipantInfo.toJSON(e) : undefined);
        }
        else {
            obj.otherParticipants = [];
        }
        message.serverVersion !== undefined &&
            (obj.serverVersion = message.serverVersion);
        if (message.iceServers) {
            obj.iceServers = message.iceServers.map((e) => e ? exports.ICEServer.toJSON(e) : undefined);
        }
        else {
            obj.iceServers = [];
        }
        message.subscriberPrimary !== undefined &&
            (obj.subscriberPrimary = message.subscriberPrimary);
        message.alternativeUrl !== undefined &&
            (obj.alternativeUrl = message.alternativeUrl);
        message.clientConfiguration !== undefined &&
            (obj.clientConfiguration = message.clientConfiguration
                ? livekit_models_1.ClientConfiguration.toJSON(message.clientConfiguration)
                : undefined);
        message.serverRegion !== undefined &&
            (obj.serverRegion = message.serverRegion);
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f;
        const message = createBaseJoinResponse();
        message.room =
            object.room !== undefined && object.room !== null
                ? livekit_models_1.Room.fromPartial(object.room)
                : undefined;
        message.participant =
            object.participant !== undefined && object.participant !== null
                ? livekit_models_1.ParticipantInfo.fromPartial(object.participant)
                : undefined;
        message.otherParticipants =
            ((_a = object.otherParticipants) === null || _a === void 0 ? void 0 : _a.map((e) => livekit_models_1.ParticipantInfo.fromPartial(e))) ||
                [];
        message.serverVersion = (_b = object.serverVersion) !== null && _b !== void 0 ? _b : "";
        message.iceServers =
            ((_c = object.iceServers) === null || _c === void 0 ? void 0 : _c.map((e) => exports.ICEServer.fromPartial(e))) || [];
        message.subscriberPrimary = (_d = object.subscriberPrimary) !== null && _d !== void 0 ? _d : false;
        message.alternativeUrl = (_e = object.alternativeUrl) !== null && _e !== void 0 ? _e : "";
        message.clientConfiguration =
            object.clientConfiguration !== undefined &&
                object.clientConfiguration !== null
                ? livekit_models_1.ClientConfiguration.fromPartial(object.clientConfiguration)
                : undefined;
        message.serverRegion = (_f = object.serverRegion) !== null && _f !== void 0 ? _f : "";
        return message;
    },
};
function createBaseTrackPublishedResponse() {
    return { cid: "", track: undefined };
}
exports.TrackPublishedResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.cid !== "") {
            writer.uint32(10).string(message.cid);
        }
        if (message.track !== undefined) {
            livekit_models_1.TrackInfo.encode(message.track, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTrackPublishedResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.cid = reader.string();
                    break;
                case 2:
                    message.track = livekit_models_1.TrackInfo.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            cid: isSet(object.cid) ? String(object.cid) : "",
            track: isSet(object.track) ? livekit_models_1.TrackInfo.fromJSON(object.track) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.cid !== undefined && (obj.cid = message.cid);
        message.track !== undefined &&
            (obj.track = message.track ? livekit_models_1.TrackInfo.toJSON(message.track) : undefined);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseTrackPublishedResponse();
        message.cid = (_a = object.cid) !== null && _a !== void 0 ? _a : "";
        message.track =
            object.track !== undefined && object.track !== null
                ? livekit_models_1.TrackInfo.fromPartial(object.track)
                : undefined;
        return message;
    },
};
function createBaseTrackUnpublishedResponse() {
    return { trackSid: "" };
}
exports.TrackUnpublishedResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.trackSid !== "") {
            writer.uint32(10).string(message.trackSid);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTrackUnpublishedResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.trackSid = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            trackSid: isSet(object.trackSid) ? String(object.trackSid) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.trackSid !== undefined && (obj.trackSid = message.trackSid);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseTrackUnpublishedResponse();
        message.trackSid = (_a = object.trackSid) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
function createBaseSessionDescription() {
    return { type: "", sdp: "" };
}
exports.SessionDescription = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.type !== "") {
            writer.uint32(10).string(message.type);
        }
        if (message.sdp !== "") {
            writer.uint32(18).string(message.sdp);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSessionDescription();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.type = reader.string();
                    break;
                case 2:
                    message.sdp = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            type: isSet(object.type) ? String(object.type) : "",
            sdp: isSet(object.sdp) ? String(object.sdp) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.type !== undefined && (obj.type = message.type);
        message.sdp !== undefined && (obj.sdp = message.sdp);
        return obj;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseSessionDescription();
        message.type = (_a = object.type) !== null && _a !== void 0 ? _a : "";
        message.sdp = (_b = object.sdp) !== null && _b !== void 0 ? _b : "";
        return message;
    },
};
function createBaseParticipantUpdate() {
    return { participants: [] };
}
exports.ParticipantUpdate = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.participants) {
            livekit_models_1.ParticipantInfo.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParticipantUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.participants.push(livekit_models_1.ParticipantInfo.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            participants: Array.isArray(object === null || object === void 0 ? void 0 : object.participants)
                ? object.participants.map((e) => livekit_models_1.ParticipantInfo.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.participants) {
            obj.participants = message.participants.map((e) => e ? livekit_models_1.ParticipantInfo.toJSON(e) : undefined);
        }
        else {
            obj.participants = [];
        }
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseParticipantUpdate();
        message.participants =
            ((_a = object.participants) === null || _a === void 0 ? void 0 : _a.map((e) => livekit_models_1.ParticipantInfo.fromPartial(e))) || [];
        return message;
    },
};
function createBaseUpdateSubscription() {
    return { trackSids: [], subscribe: false, participantTracks: [] };
}
exports.UpdateSubscription = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.trackSids) {
            writer.uint32(10).string(v);
        }
        if (message.subscribe === true) {
            writer.uint32(16).bool(message.subscribe);
        }
        for (const v of message.participantTracks) {
            livekit_models_1.ParticipantTracks.encode(v, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseUpdateSubscription();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.trackSids.push(reader.string());
                    break;
                case 2:
                    message.subscribe = reader.bool();
                    break;
                case 3:
                    message.participantTracks.push(livekit_models_1.ParticipantTracks.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            trackSids: Array.isArray(object === null || object === void 0 ? void 0 : object.trackSids)
                ? object.trackSids.map((e) => String(e))
                : [],
            subscribe: isSet(object.subscribe) ? Boolean(object.subscribe) : false,
            participantTracks: Array.isArray(object === null || object === void 0 ? void 0 : object.participantTracks)
                ? object.participantTracks.map((e) => livekit_models_1.ParticipantTracks.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.trackSids) {
            obj.trackSids = message.trackSids.map((e) => e);
        }
        else {
            obj.trackSids = [];
        }
        message.subscribe !== undefined && (obj.subscribe = message.subscribe);
        if (message.participantTracks) {
            obj.participantTracks = message.participantTracks.map((e) => e ? livekit_models_1.ParticipantTracks.toJSON(e) : undefined);
        }
        else {
            obj.participantTracks = [];
        }
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseUpdateSubscription();
        message.trackSids = ((_a = object.trackSids) === null || _a === void 0 ? void 0 : _a.map((e) => e)) || [];
        message.subscribe = (_b = object.subscribe) !== null && _b !== void 0 ? _b : false;
        message.participantTracks =
            ((_c = object.participantTracks) === null || _c === void 0 ? void 0 : _c.map((e) => livekit_models_1.ParticipantTracks.fromPartial(e))) ||
                [];
        return message;
    },
};
function createBaseUpdateTrackSettings() {
    return { trackSids: [], disabled: false, quality: 0, width: 0, height: 0 };
}
exports.UpdateTrackSettings = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.trackSids) {
            writer.uint32(10).string(v);
        }
        if (message.disabled === true) {
            writer.uint32(24).bool(message.disabled);
        }
        if (message.quality !== 0) {
            writer.uint32(32).int32(message.quality);
        }
        if (message.width !== 0) {
            writer.uint32(40).uint32(message.width);
        }
        if (message.height !== 0) {
            writer.uint32(48).uint32(message.height);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseUpdateTrackSettings();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.trackSids.push(reader.string());
                    break;
                case 3:
                    message.disabled = reader.bool();
                    break;
                case 4:
                    message.quality = reader.int32();
                    break;
                case 5:
                    message.width = reader.uint32();
                    break;
                case 6:
                    message.height = reader.uint32();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            trackSids: Array.isArray(object === null || object === void 0 ? void 0 : object.trackSids)
                ? object.trackSids.map((e) => String(e))
                : [],
            disabled: isSet(object.disabled) ? Boolean(object.disabled) : false,
            quality: isSet(object.quality) ? livekit_models_1.videoQualityFromJSON(object.quality) : 0,
            width: isSet(object.width) ? Number(object.width) : 0,
            height: isSet(object.height) ? Number(object.height) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.trackSids) {
            obj.trackSids = message.trackSids.map((e) => e);
        }
        else {
            obj.trackSids = [];
        }
        message.disabled !== undefined && (obj.disabled = message.disabled);
        message.quality !== undefined &&
            (obj.quality = livekit_models_1.videoQualityToJSON(message.quality));
        message.width !== undefined && (obj.width = Math.round(message.width));
        message.height !== undefined && (obj.height = Math.round(message.height));
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e;
        const message = createBaseUpdateTrackSettings();
        message.trackSids = ((_a = object.trackSids) === null || _a === void 0 ? void 0 : _a.map((e) => e)) || [];
        message.disabled = (_b = object.disabled) !== null && _b !== void 0 ? _b : false;
        message.quality = (_c = object.quality) !== null && _c !== void 0 ? _c : 0;
        message.width = (_d = object.width) !== null && _d !== void 0 ? _d : 0;
        message.height = (_e = object.height) !== null && _e !== void 0 ? _e : 0;
        return message;
    },
};
function createBaseLeaveRequest() {
    return { canReconnect: false };
}
exports.LeaveRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.canReconnect === true) {
            writer.uint32(8).bool(message.canReconnect);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseLeaveRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.canReconnect = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            canReconnect: isSet(object.canReconnect)
                ? Boolean(object.canReconnect)
                : false,
        };
    },
    toJSON(message) {
        const obj = {};
        message.canReconnect !== undefined &&
            (obj.canReconnect = message.canReconnect);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseLeaveRequest();
        message.canReconnect = (_a = object.canReconnect) !== null && _a !== void 0 ? _a : false;
        return message;
    },
};
function createBaseUpdateVideoLayers() {
    return { trackSid: "", layers: [] };
}
exports.UpdateVideoLayers = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.trackSid !== "") {
            writer.uint32(10).string(message.trackSid);
        }
        for (const v of message.layers) {
            livekit_models_1.VideoLayer.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseUpdateVideoLayers();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.trackSid = reader.string();
                    break;
                case 2:
                    message.layers.push(livekit_models_1.VideoLayer.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            trackSid: isSet(object.trackSid) ? String(object.trackSid) : "",
            layers: Array.isArray(object === null || object === void 0 ? void 0 : object.layers)
                ? object.layers.map((e) => livekit_models_1.VideoLayer.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        message.trackSid !== undefined && (obj.trackSid = message.trackSid);
        if (message.layers) {
            obj.layers = message.layers.map((e) => e ? livekit_models_1.VideoLayer.toJSON(e) : undefined);
        }
        else {
            obj.layers = [];
        }
        return obj;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseUpdateVideoLayers();
        message.trackSid = (_a = object.trackSid) !== null && _a !== void 0 ? _a : "";
        message.layers = ((_b = object.layers) === null || _b === void 0 ? void 0 : _b.map((e) => livekit_models_1.VideoLayer.fromPartial(e))) || [];
        return message;
    },
};
function createBaseICEServer() {
    return { urls: [], username: "", credential: "" };
}
exports.ICEServer = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.urls) {
            writer.uint32(10).string(v);
        }
        if (message.username !== "") {
            writer.uint32(18).string(message.username);
        }
        if (message.credential !== "") {
            writer.uint32(26).string(message.credential);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseICEServer();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.urls.push(reader.string());
                    break;
                case 2:
                    message.username = reader.string();
                    break;
                case 3:
                    message.credential = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            urls: Array.isArray(object === null || object === void 0 ? void 0 : object.urls)
                ? object.urls.map((e) => String(e))
                : [],
            username: isSet(object.username) ? String(object.username) : "",
            credential: isSet(object.credential) ? String(object.credential) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.urls) {
            obj.urls = message.urls.map((e) => e);
        }
        else {
            obj.urls = [];
        }
        message.username !== undefined && (obj.username = message.username);
        message.credential !== undefined && (obj.credential = message.credential);
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseICEServer();
        message.urls = ((_a = object.urls) === null || _a === void 0 ? void 0 : _a.map((e) => e)) || [];
        message.username = (_b = object.username) !== null && _b !== void 0 ? _b : "";
        message.credential = (_c = object.credential) !== null && _c !== void 0 ? _c : "";
        return message;
    },
};
function createBaseSpeakersChanged() {
    return { speakers: [] };
}
exports.SpeakersChanged = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.speakers) {
            livekit_models_1.SpeakerInfo.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSpeakersChanged();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.speakers.push(livekit_models_1.SpeakerInfo.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            speakers: Array.isArray(object === null || object === void 0 ? void 0 : object.speakers)
                ? object.speakers.map((e) => livekit_models_1.SpeakerInfo.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.speakers) {
            obj.speakers = message.speakers.map((e) => e ? livekit_models_1.SpeakerInfo.toJSON(e) : undefined);
        }
        else {
            obj.speakers = [];
        }
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseSpeakersChanged();
        message.speakers =
            ((_a = object.speakers) === null || _a === void 0 ? void 0 : _a.map((e) => livekit_models_1.SpeakerInfo.fromPartial(e))) || [];
        return message;
    },
};
function createBaseRoomUpdate() {
    return { room: undefined };
}
exports.RoomUpdate = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.room !== undefined) {
            livekit_models_1.Room.encode(message.room, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRoomUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.room = livekit_models_1.Room.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            room: isSet(object.room) ? livekit_models_1.Room.fromJSON(object.room) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.room !== undefined &&
            (obj.room = message.room ? livekit_models_1.Room.toJSON(message.room) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseRoomUpdate();
        message.room =
            object.room !== undefined && object.room !== null
                ? livekit_models_1.Room.fromPartial(object.room)
                : undefined;
        return message;
    },
};
function createBaseConnectionQualityInfo() {
    return { participantSid: "", quality: 0, score: 0 };
}
exports.ConnectionQualityInfo = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.participantSid !== "") {
            writer.uint32(10).string(message.participantSid);
        }
        if (message.quality !== 0) {
            writer.uint32(16).int32(message.quality);
        }
        if (message.score !== 0) {
            writer.uint32(29).float(message.score);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConnectionQualityInfo();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.participantSid = reader.string();
                    break;
                case 2:
                    message.quality = reader.int32();
                    break;
                case 3:
                    message.score = reader.float();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            participantSid: isSet(object.participantSid)
                ? String(object.participantSid)
                : "",
            quality: isSet(object.quality)
                ? livekit_models_1.connectionQualityFromJSON(object.quality)
                : 0,
            score: isSet(object.score) ? Number(object.score) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        message.participantSid !== undefined &&
            (obj.participantSid = message.participantSid);
        message.quality !== undefined &&
            (obj.quality = livekit_models_1.connectionQualityToJSON(message.quality));
        message.score !== undefined && (obj.score = message.score);
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseConnectionQualityInfo();
        message.participantSid = (_a = object.participantSid) !== null && _a !== void 0 ? _a : "";
        message.quality = (_b = object.quality) !== null && _b !== void 0 ? _b : 0;
        message.score = (_c = object.score) !== null && _c !== void 0 ? _c : 0;
        return message;
    },
};
function createBaseConnectionQualityUpdate() {
    return { updates: [] };
}
exports.ConnectionQualityUpdate = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.updates) {
            exports.ConnectionQualityInfo.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConnectionQualityUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.updates.push(exports.ConnectionQualityInfo.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            updates: Array.isArray(object === null || object === void 0 ? void 0 : object.updates)
                ? object.updates.map((e) => exports.ConnectionQualityInfo.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.updates) {
            obj.updates = message.updates.map((e) => e ? exports.ConnectionQualityInfo.toJSON(e) : undefined);
        }
        else {
            obj.updates = [];
        }
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseConnectionQualityUpdate();
        message.updates =
            ((_a = object.updates) === null || _a === void 0 ? void 0 : _a.map((e) => exports.ConnectionQualityInfo.fromPartial(e))) || [];
        return message;
    },
};
function createBaseStreamStateInfo() {
    return { participantSid: "", trackSid: "", state: 0 };
}
exports.StreamStateInfo = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.participantSid !== "") {
            writer.uint32(10).string(message.participantSid);
        }
        if (message.trackSid !== "") {
            writer.uint32(18).string(message.trackSid);
        }
        if (message.state !== 0) {
            writer.uint32(24).int32(message.state);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStreamStateInfo();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.participantSid = reader.string();
                    break;
                case 2:
                    message.trackSid = reader.string();
                    break;
                case 3:
                    message.state = reader.int32();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            participantSid: isSet(object.participantSid)
                ? String(object.participantSid)
                : "",
            trackSid: isSet(object.trackSid) ? String(object.trackSid) : "",
            state: isSet(object.state) ? streamStateFromJSON(object.state) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        message.participantSid !== undefined &&
            (obj.participantSid = message.participantSid);
        message.trackSid !== undefined && (obj.trackSid = message.trackSid);
        message.state !== undefined &&
            (obj.state = streamStateToJSON(message.state));
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseStreamStateInfo();
        message.participantSid = (_a = object.participantSid) !== null && _a !== void 0 ? _a : "";
        message.trackSid = (_b = object.trackSid) !== null && _b !== void 0 ? _b : "";
        message.state = (_c = object.state) !== null && _c !== void 0 ? _c : 0;
        return message;
    },
};
function createBaseStreamStateUpdate() {
    return { streamStates: [] };
}
exports.StreamStateUpdate = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.streamStates) {
            exports.StreamStateInfo.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStreamStateUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.streamStates.push(exports.StreamStateInfo.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            streamStates: Array.isArray(object === null || object === void 0 ? void 0 : object.streamStates)
                ? object.streamStates.map((e) => exports.StreamStateInfo.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.streamStates) {
            obj.streamStates = message.streamStates.map((e) => e ? exports.StreamStateInfo.toJSON(e) : undefined);
        }
        else {
            obj.streamStates = [];
        }
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseStreamStateUpdate();
        message.streamStates =
            ((_a = object.streamStates) === null || _a === void 0 ? void 0 : _a.map((e) => exports.StreamStateInfo.fromPartial(e))) || [];
        return message;
    },
};
function createBaseSubscribedQuality() {
    return { quality: 0, enabled: false };
}
exports.SubscribedQuality = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.quality !== 0) {
            writer.uint32(8).int32(message.quality);
        }
        if (message.enabled === true) {
            writer.uint32(16).bool(message.enabled);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSubscribedQuality();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.quality = reader.int32();
                    break;
                case 2:
                    message.enabled = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            quality: isSet(object.quality) ? livekit_models_1.videoQualityFromJSON(object.quality) : 0,
            enabled: isSet(object.enabled) ? Boolean(object.enabled) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        message.quality !== undefined &&
            (obj.quality = livekit_models_1.videoQualityToJSON(message.quality));
        message.enabled !== undefined && (obj.enabled = message.enabled);
        return obj;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseSubscribedQuality();
        message.quality = (_a = object.quality) !== null && _a !== void 0 ? _a : 0;
        message.enabled = (_b = object.enabled) !== null && _b !== void 0 ? _b : false;
        return message;
    },
};
function createBaseSubscribedQualityUpdate() {
    return { trackSid: "", subscribedQualities: [] };
}
exports.SubscribedQualityUpdate = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.trackSid !== "") {
            writer.uint32(10).string(message.trackSid);
        }
        for (const v of message.subscribedQualities) {
            exports.SubscribedQuality.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSubscribedQualityUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.trackSid = reader.string();
                    break;
                case 2:
                    message.subscribedQualities.push(exports.SubscribedQuality.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            trackSid: isSet(object.trackSid) ? String(object.trackSid) : "",
            subscribedQualities: Array.isArray(object === null || object === void 0 ? void 0 : object.subscribedQualities)
                ? object.subscribedQualities.map((e) => exports.SubscribedQuality.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        message.trackSid !== undefined && (obj.trackSid = message.trackSid);
        if (message.subscribedQualities) {
            obj.subscribedQualities = message.subscribedQualities.map((e) => e ? exports.SubscribedQuality.toJSON(e) : undefined);
        }
        else {
            obj.subscribedQualities = [];
        }
        return obj;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseSubscribedQualityUpdate();
        message.trackSid = (_a = object.trackSid) !== null && _a !== void 0 ? _a : "";
        message.subscribedQualities =
            ((_b = object.subscribedQualities) === null || _b === void 0 ? void 0 : _b.map((e) => exports.SubscribedQuality.fromPartial(e))) || [];
        return message;
    },
};
function createBaseTrackPermission() {
    return { participantSid: "", allTracks: false, trackSids: [] };
}
exports.TrackPermission = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.participantSid !== "") {
            writer.uint32(10).string(message.participantSid);
        }
        if (message.allTracks === true) {
            writer.uint32(16).bool(message.allTracks);
        }
        for (const v of message.trackSids) {
            writer.uint32(26).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTrackPermission();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.participantSid = reader.string();
                    break;
                case 2:
                    message.allTracks = reader.bool();
                    break;
                case 3:
                    message.trackSids.push(reader.string());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            participantSid: isSet(object.participantSid)
                ? String(object.participantSid)
                : "",
            allTracks: isSet(object.allTracks) ? Boolean(object.allTracks) : false,
            trackSids: Array.isArray(object === null || object === void 0 ? void 0 : object.trackSids)
                ? object.trackSids.map((e) => String(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        message.participantSid !== undefined &&
            (obj.participantSid = message.participantSid);
        message.allTracks !== undefined && (obj.allTracks = message.allTracks);
        if (message.trackSids) {
            obj.trackSids = message.trackSids.map((e) => e);
        }
        else {
            obj.trackSids = [];
        }
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseTrackPermission();
        message.participantSid = (_a = object.participantSid) !== null && _a !== void 0 ? _a : "";
        message.allTracks = (_b = object.allTracks) !== null && _b !== void 0 ? _b : false;
        message.trackSids = ((_c = object.trackSids) === null || _c === void 0 ? void 0 : _c.map((e) => e)) || [];
        return message;
    },
};
function createBaseSubscriptionPermission() {
    return { allParticipants: false, trackPermissions: [] };
}
exports.SubscriptionPermission = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.allParticipants === true) {
            writer.uint32(8).bool(message.allParticipants);
        }
        for (const v of message.trackPermissions) {
            exports.TrackPermission.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSubscriptionPermission();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.allParticipants = reader.bool();
                    break;
                case 2:
                    message.trackPermissions.push(exports.TrackPermission.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            allParticipants: isSet(object.allParticipants)
                ? Boolean(object.allParticipants)
                : false,
            trackPermissions: Array.isArray(object === null || object === void 0 ? void 0 : object.trackPermissions)
                ? object.trackPermissions.map((e) => exports.TrackPermission.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        message.allParticipants !== undefined &&
            (obj.allParticipants = message.allParticipants);
        if (message.trackPermissions) {
            obj.trackPermissions = message.trackPermissions.map((e) => e ? exports.TrackPermission.toJSON(e) : undefined);
        }
        else {
            obj.trackPermissions = [];
        }
        return obj;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseSubscriptionPermission();
        message.allParticipants = (_a = object.allParticipants) !== null && _a !== void 0 ? _a : false;
        message.trackPermissions =
            ((_b = object.trackPermissions) === null || _b === void 0 ? void 0 : _b.map((e) => exports.TrackPermission.fromPartial(e))) || [];
        return message;
    },
};
function createBaseSubscriptionPermissionUpdate() {
    return { participantSid: "", trackSid: "", allowed: false };
}
exports.SubscriptionPermissionUpdate = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.participantSid !== "") {
            writer.uint32(10).string(message.participantSid);
        }
        if (message.trackSid !== "") {
            writer.uint32(18).string(message.trackSid);
        }
        if (message.allowed === true) {
            writer.uint32(24).bool(message.allowed);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSubscriptionPermissionUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.participantSid = reader.string();
                    break;
                case 2:
                    message.trackSid = reader.string();
                    break;
                case 3:
                    message.allowed = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            participantSid: isSet(object.participantSid)
                ? String(object.participantSid)
                : "",
            trackSid: isSet(object.trackSid) ? String(object.trackSid) : "",
            allowed: isSet(object.allowed) ? Boolean(object.allowed) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        message.participantSid !== undefined &&
            (obj.participantSid = message.participantSid);
        message.trackSid !== undefined && (obj.trackSid = message.trackSid);
        message.allowed !== undefined && (obj.allowed = message.allowed);
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseSubscriptionPermissionUpdate();
        message.participantSid = (_a = object.participantSid) !== null && _a !== void 0 ? _a : "";
        message.trackSid = (_b = object.trackSid) !== null && _b !== void 0 ? _b : "";
        message.allowed = (_c = object.allowed) !== null && _c !== void 0 ? _c : false;
        return message;
    },
};
function createBaseSyncState() {
    return {
        answer: undefined,
        subscription: undefined,
        publishTracks: [],
        dataChannels: [],
    };
}
exports.SyncState = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.answer !== undefined) {
            exports.SessionDescription.encode(message.answer, writer.uint32(10).fork()).ldelim();
        }
        if (message.subscription !== undefined) {
            exports.UpdateSubscription.encode(message.subscription, writer.uint32(18).fork()).ldelim();
        }
        for (const v of message.publishTracks) {
            exports.TrackPublishedResponse.encode(v, writer.uint32(26).fork()).ldelim();
        }
        for (const v of message.dataChannels) {
            exports.DataChannelInfo.encode(v, writer.uint32(34).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSyncState();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.answer = exports.SessionDescription.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.subscription = exports.UpdateSubscription.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.publishTracks.push(exports.TrackPublishedResponse.decode(reader, reader.uint32()));
                    break;
                case 4:
                    message.dataChannels.push(exports.DataChannelInfo.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            answer: isSet(object.answer)
                ? exports.SessionDescription.fromJSON(object.answer)
                : undefined,
            subscription: isSet(object.subscription)
                ? exports.UpdateSubscription.fromJSON(object.subscription)
                : undefined,
            publishTracks: Array.isArray(object === null || object === void 0 ? void 0 : object.publishTracks)
                ? object.publishTracks.map((e) => exports.TrackPublishedResponse.fromJSON(e))
                : [],
            dataChannels: Array.isArray(object === null || object === void 0 ? void 0 : object.dataChannels)
                ? object.dataChannels.map((e) => exports.DataChannelInfo.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        message.answer !== undefined &&
            (obj.answer = message.answer
                ? exports.SessionDescription.toJSON(message.answer)
                : undefined);
        message.subscription !== undefined &&
            (obj.subscription = message.subscription
                ? exports.UpdateSubscription.toJSON(message.subscription)
                : undefined);
        if (message.publishTracks) {
            obj.publishTracks = message.publishTracks.map((e) => e ? exports.TrackPublishedResponse.toJSON(e) : undefined);
        }
        else {
            obj.publishTracks = [];
        }
        if (message.dataChannels) {
            obj.dataChannels = message.dataChannels.map((e) => e ? exports.DataChannelInfo.toJSON(e) : undefined);
        }
        else {
            obj.dataChannels = [];
        }
        return obj;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseSyncState();
        message.answer =
            object.answer !== undefined && object.answer !== null
                ? exports.SessionDescription.fromPartial(object.answer)
                : undefined;
        message.subscription =
            object.subscription !== undefined && object.subscription !== null
                ? exports.UpdateSubscription.fromPartial(object.subscription)
                : undefined;
        message.publishTracks =
            ((_a = object.publishTracks) === null || _a === void 0 ? void 0 : _a.map((e) => exports.TrackPublishedResponse.fromPartial(e))) ||
                [];
        message.dataChannels =
            ((_b = object.dataChannels) === null || _b === void 0 ? void 0 : _b.map((e) => exports.DataChannelInfo.fromPartial(e))) || [];
        return message;
    },
};
function createBaseDataChannelInfo() {
    return { label: "", id: 0 };
}
exports.DataChannelInfo = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.label !== "") {
            writer.uint32(10).string(message.label);
        }
        if (message.id !== 0) {
            writer.uint32(16).uint32(message.id);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDataChannelInfo();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.label = reader.string();
                    break;
                case 2:
                    message.id = reader.uint32();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            label: isSet(object.label) ? String(object.label) : "",
            id: isSet(object.id) ? Number(object.id) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        message.label !== undefined && (obj.label = message.label);
        message.id !== undefined && (obj.id = Math.round(message.id));
        return obj;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseDataChannelInfo();
        message.label = (_a = object.label) !== null && _a !== void 0 ? _a : "";
        message.id = (_b = object.id) !== null && _b !== void 0 ? _b : 0;
        return message;
    },
};
function createBaseSimulateScenario() {
    return {
        speakerUpdate: undefined,
        nodeFailure: undefined,
        migration: undefined,
        serverLeave: undefined,
    };
}
exports.SimulateScenario = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.speakerUpdate !== undefined) {
            writer.uint32(8).int32(message.speakerUpdate);
        }
        if (message.nodeFailure !== undefined) {
            writer.uint32(16).bool(message.nodeFailure);
        }
        if (message.migration !== undefined) {
            writer.uint32(24).bool(message.migration);
        }
        if (message.serverLeave !== undefined) {
            writer.uint32(32).bool(message.serverLeave);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSimulateScenario();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.speakerUpdate = reader.int32();
                    break;
                case 2:
                    message.nodeFailure = reader.bool();
                    break;
                case 3:
                    message.migration = reader.bool();
                    break;
                case 4:
                    message.serverLeave = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            speakerUpdate: isSet(object.speakerUpdate)
                ? Number(object.speakerUpdate)
                : undefined,
            nodeFailure: isSet(object.nodeFailure)
                ? Boolean(object.nodeFailure)
                : undefined,
            migration: isSet(object.migration)
                ? Boolean(object.migration)
                : undefined,
            serverLeave: isSet(object.serverLeave)
                ? Boolean(object.serverLeave)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.speakerUpdate !== undefined &&
            (obj.speakerUpdate = Math.round(message.speakerUpdate));
        message.nodeFailure !== undefined &&
            (obj.nodeFailure = message.nodeFailure);
        message.migration !== undefined && (obj.migration = message.migration);
        message.serverLeave !== undefined &&
            (obj.serverLeave = message.serverLeave);
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c, _d;
        const message = createBaseSimulateScenario();
        message.speakerUpdate = (_a = object.speakerUpdate) !== null && _a !== void 0 ? _a : undefined;
        message.nodeFailure = (_b = object.nodeFailure) !== null && _b !== void 0 ? _b : undefined;
        message.migration = (_c = object.migration) !== null && _c !== void 0 ? _c : undefined;
        message.serverLeave = (_d = object.serverLeave) !== null && _d !== void 0 ? _d : undefined;
        return message;
    },
};
if (_m0.util.Long !== long_1.default) {
    _m0.util.Long = long_1.default;
    _m0.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
